rfc-135-updating-routes-in-router.md

```diff
@@ -0,0 +1,44 @@
+
+## Proposal
+
+Router will keep its in-memory cache of routes up-to-date via a two stage process:
+
+  1. Router will perform a simple query against MongoDB returning the state of the database with the current set of routes; ideally this will be a query which contains a checksum or timestamp of last update which can be used by Router instances to check whether their local copy of routes is up-to-date.
+  2. If the check against MongoDB concludes that Router needs to update its in-memory cache of routes, then Router will self-action a reload of all routes, using the same mechanism as is currently does when it receives a call to its `/reload` endpoint.
+
+The polling of changes by Router against MongoDB will take place every few seconds in order to ensure that routes are updated quickly as and when they change, which keeps in line with the current approach of updating routes when comparing how fast route updates propagate through to Router instances. Individual Router instances will drift slightly out of sync with each other, but because we're polling every few seconds and updating as soon as we see changes there won't be any noticeable changes.
+
+This proposal is relatively simple to implement compared to some of the others we have considered and will allow us to move at pace in decoupling Router API and Router. It also puts us in a good place to iterate the solution further by using the change streams feature in future, whether that's offered by MongoDB, DocumentDB or any other database technology that we decide to use for storing routes.
+
+### Possible alternatives
+
+There were a number of approaches considered to tackle the problem presented in this RFC and [these approaches have been documented in detail](https://docs.google.com/document/d/1gGRWTmhfcfU1jfBDWYMiDa-6G64tRNdPFI3BLcbCxWc/edit#). Still, it is worth a brief summary of the main approaches discussed for completeness.
+
+#### Removing Router API entirely
+
+The first option available to us was to remove Router API entirely, which has been [discussed previously](https://github.com/alphagov/govuk-rfcs/pull/72) and for which there also exists a [tech debt card](https://trello.com/c/MOuw5ke0). Removal of Router API from the stack would be a good refactoring for GOV.UK to remove complexity, as Content Store already calls Router API, so why not remove this call and just have Content Store update routes directly.
+
+The downside to this approach is that it doesn't solve the underlying problem of this RFC, namely that as part of moving apps to ECS and within AWS App Mesh, it's not going to be possible to talk to individual instances directly. Therefore the problem moves to Content Store not being able to talk to Router instances rather than Router API not being able to talk to Router instances.
```

Comment:
I think talking to individual instances should be possible - each router ECS task has an IP address. We just don't want to implement this, right? Perhaps we could say something like "we'd need to implement a way for Router API to reach Router instances in ECS, which we think is not a good approach given the incidents".

You could also add something about how removing Router API might not make things simpler, since it's quite a simple service app with essentially one endpoint. Moving Router API's logic into Content Store might introduce complexity (and perhaps unforeseen scaling issues).

Reply:
I was under the impression that this _wouldn't_ actually be possible with ECS _and_ App Mesh, perhaps @richardTowers can clarify?

As for removing Router API and moving the logic into Content Store, I think the RFC mentioned in this RFC which discussed removing Router API covers that topic in good detail; I think this move _would_ introduce some additional complexity, but given that it's Content Store already calling Router API, I wouldn't have thought this might introduce scaling issues? Happy to be proven wrong here.

Reply:
It _may_ be possible, I'm not sure I know enough about the details of how AppMesh does it's "load balancing" to be sure.

In my current mental model of how it works I think we could do:

* make a DNS request for `router.mesh.govuk-internal.digital`
* this will return the IP addresses of all the router instances
* contact each IP individually 🎉 

But ... is that what's really happening? Or is Envoy more involved than I suspect it is? Will it refuse to let us talk to a raw IP that it doesn't know about?

Even if this would work, it feels verrrryy hacky.

Maybe:

> The downside is that this doesn't solve the underlying problem of this RFC. When apps are running in ECS / AppMesh, it is going to be difficult to contact individual instances of router without doing something hacky. Removing Router API just moves the problem to Content Store not being able to talk to Router instances, it doesn't make it go away.

Reply:
Yeah I agree, we don't want to do this, [regardless of whether it's possible](https://www.youtube.com/watch?v=kY-pUxKQMUE). I think the fact that we could achieve it - but with hacks, would be good to note in the problem section.

An aside: we should really find out what Envoy does with requests 😬 . Also, the AWS APIs are an alternative to DNS to get a list of IP address. E.g. `aws ecs describe-tasks` returns the `privateIpv4Address` for a set of Fargate tasks.

Reply:
I have added Richard's points to the _problem_ section of this RFC as Bill suggests; take a look and let us know your thoughts.

Reply:
The problem section looks good, thanks @karlbaker02.

Reply:
(Sorry for resurrecting old thread but it seemed like the most relevant place for this.)

Having Router API fetch the list of Router instances via DNS would have been a tempting solution IMO (replace reading the instances file with a DNS query and re-query+retry with backoff when contacting an instance fails), but it turns out that an arbitrary limitation in AWS makes it unviable: Amazon's cluster DNS (Cloud Map + ECS) only returns [8 instances maximum](https://docs.aws.amazon.com/cloud-map/latest/api/API_DnsConfig.html#cloudmap-Type-DnsConfig-RoutingPolicy). Amazon has an API which returns the full set of instances, but tying Router API to a proprietary cloud vendor API seems like a step too far (and having Router API watch the database is a more general solution as it removes unnecessary coupling between Router API and Router and should be less complex overall).